File: android/src/main/kotlin/sncf/connect/tech/flutter_eco_mode/listener/PowerModeListener.kt
package sncf.connect.tech.flutter_eco_mode.listener

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.EventChannel
import sncf.connect.tech.flutter_eco_mode.BatteryModeStreamHandler
import sncf.connect.tech.flutter_eco_mode.EcoBatteryManager
import sncf.connect.tech.flutter_eco_mode.MessagesPigeonMethodCodec
import sncf.connect.tech.flutter_eco_mode.PigeonEventSink

class PowerModeListener(
    private val ecoBatteryManager: EcoBatteryManager,
) : BatteryModeStreamHandler(), DisposableStreamListener {
    private var lowPowerModeEventSink: PigeonEventSink<Boolean>? = null
    private var powerSavingReceiver: BroadcastReceiver? = null
    private var lastPowerSaveMode: Boolean? = null

    override fun onListen(p0: Any?, sink: PigeonEventSink<Boolean>) {
        lowPowerModeEventSink = sink

        powerSavingReceiver = object : BroadcastReceiver() {
            override fun onReceive(context: Context, intent: Intent?) {
                sendPowerSaveUpdate(ecoBatteryManager.isLowPowerMode())
            }
        }

        ecoBatteryManager.registerPowerModeReceiver(powerSavingReceiver)
        sendPowerSaveUpdate(ecoBatteryManager.isLowPowerMode())
    }

    override fun onCancel(p0: Any?) {
        cleanUp()
    }

    override fun register(binaryMessenger: BinaryMessenger) = register(binaryMessenger, this)

    override fun dispose(binaryMessenger: BinaryMessenger) {
        cleanUp()

        EventChannel(binaryMessenger, EcoModeEventChannels.BATTERY_MODE, MessagesPigeonMethodCodec).setStreamHandler(null)
    }

    private fun cleanUp() {
        powerSavingReceiver?.let { ecoBatteryManager.unregisterReceiver(it) }
        powerSavingReceiver = null
        lowPowerModeEventSink = null
        lastPowerSaveMode = null
    }

    private fun sendPowerSaveUpdate(isPowerSaveMode: Boolean) {
        if (isPowerSaveMode != lastPowerSaveMode) {
            lastPowerSaveMode = isPowerSaveMode
            lowPowerModeEventSink?.success(isPowerSaveMode)
        }
    }
}

File: ios/flutter_eco_mode/Sources/flutter-eco-mode/listeners/PowerModeListener.swift
//
//  PowerModeListener.swift
//  flutter_eco_mode
//
//  Created by CHOUPAULT Alexis on 27/02/2026.
//

import Flutter

class PowerModeListener: BatteryModeStreamHandler, DisposableStreamListener {
    private var eventSink: PigeonEventSink<Bool>?
    private var lastSentPowerMode: Bool?
    
    // MARK: - BatteryModeStreamHandler implementation

    override func onListen(withArguments arguments: Any?, sink: PigeonEventSink<Bool>) {
        self.eventSink = sink
        
        sendUpdate()
        
        NotificationCenter.default.addObserver(
            self,
            selector: #selector(lowPowerModeChanged),
            name: Notification.Name.NSProcessInfoPowerStateDidChange,
            object: nil
        )
    }
    
    override func onCancel(withArguments arguments: Any?) {
        cleanUp()
    }
    
    // MARK: - DisposableStreamListener implementation
    
    func register(binaryMessenger: FlutterBinaryMessenger) {
        BatteryModeStreamHandler.register(with: binaryMessenger, streamHandler: self)
    }
    
    func dispose(binaryMessenger: FlutterBinaryMessenger) {
        cleanUp()
        
        FlutterEventChannel(name: EcoModeEventChannels.batteryMode, binaryMessenger: binaryMessenger, codec: messagesPigeonMethodCodec).setStreamHandler(nil)
    }
    
    // MARK: - Private methods

    @objc private func lowPowerModeChanged() {
        sendUpdate()
    }

    private func sendUpdate() {
        let isLowPower = EcoBatteryManager.shared.isLowPowerMode()
        
        guard isLowPower != lastSentPowerMode else { return }
        
        lastSentPowerMode = isLowPower
        
        DispatchQueue.main.async { [weak self] in
            self?.eventSink?.success(isLowPower)
        }
    }

    private func cleanUp() {
        NotificationCenter.default.removeObserver(self)
        eventSink = nil
        lastSentPowerMode = nil
    }
}
